test/integration/yurttunnel_test.go

```diff
@@ -0,0 +1,208 @@
+		},
+		Addr:    fmt.Sprintf(":%d", DummyServerPort),
+		Handler: m,
+	}
+
+	klog.Info("[TEST] dummy-server is listening on :9515")
+	if err := s.ListenAndServeTLS("", ""); err != nil {
+		t.Fatalf("the dummy-server failed: %v", err)
+	}
+}
+
+func startDummyClient(t *testing.T, wg *sync.WaitGroup) {
+	defer wg.Done()
+	c := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				Certificates: []tls.Certificate{
+					genCert(t, GeneircCertFile, GenericKeyFile),
+				},
+				RootCAs: genCAPool(t, RootCAFile),
+			},
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("tcp", "127.0.0.1:9516")
+			},
+		},
+	}
+
+	r, err := http.NewRequest(http.MethodGet, "https://127.0.0.1:9515/"+HTTPPath, nil)
+	if err != nil {
+		t.Fatalf("fail to generate http request: %v", err)
+	}
+
+	rep, err := c.Do(r)
+	if err != nil {
+		t.Fatalf("fail to send request to the server: %v", err)
+	}
+	defer rep.Body.Close()
+
+	bodyByts, err := ioutil.ReadAll(rep.Body)
+	if err != nil {
+		t.Fatalf("fail to read from the response body: %v", err)
+	}
+	if string(bodyByts) != ResponseForRegularRequest {
+		t.Fatalf("invalid response content, expect: Dummy Server, got: %s", string(bodyByts))
+	}
+	klog.Info("[TEST] successfully send request to the Dummy server")
+}
+
+func startTunnelServer(t *testing.T) {
+	tlsCfg := tls.Config{
+		Certificates: []tls.Certificate{
+			genCert(t, GeneircCertFile, GenericKeyFile),
+		},
+		ClientCAs: genCAPool(t, RootCAFile),
+	}
+	wrappers := hw.HandlerWrappers([]hw.Middleware{
+		tr.NewTraceReqMiddleware(),
+	})
+
+	_, err := os.Stat(InterceptorServerUDSFile)
+	if !os.IsNotExist(err) {
+		os.Remove(InterceptorServerUDSFile)
+	}
+
+	tunnelServer := ts.NewTunnelServer(
+		false,                                /*egressSelectorEnabled*/
+		InterceptorServerUDSFile,             /* interceptorServerUDSFile*/
+		fmt.Sprintf(":%d", ServerMasterPort), /* serverMasterAddr */
+		fmt.Sprintf(":%d", ServerMasterInsecurePort), /* serverMasterInsecureAddr */
+		fmt.Sprintf(":%d", ServerAgentPort),          /* serverAgentAddr */
+		1,                                            /* serverCount */
+		&tlsCfg,                                      /* tlsCfg */
+		wrappers,                                     /* hw.HandlerWrappers */
+		string(anpserver.ProxyStrategyDestHost),      /* proxyStrategy */
+	)
+	tunnelServer.Run()
+	klog.Info("[TEST] Yurttunnel Server is running")
+}
+
+func startTunnelAgent(t *testing.T) {
+	tlsCfg := tls.Config{
+		Certificates: []tls.Certificate{
+			genCert(t, GeneircCertFile, GenericKeyFile),
+		},
+		RootCAs:    genCAPool(t, RootCAFile),
+		ServerName: "127.0.0.1",
+	}
+	tunnelAgent := ta.NewTunnelAgent(
+		&tlsCfg,                             /* tlsCfg */
+		fmt.Sprintf(":%d", ServerAgentPort), /* tunnelServerAddr */
+		"dummy-agent",                       /* nodeName */
+		"127.0.0.1",                         /* agentIdentifiers */
```

Comment:
agentIdentifiers should be set with "ipv4=127.0.0.1", not only ip address.
and i think that agentIdentifiers definition is confusing, it's not so easy for user to setup. maybe we need to add some verification or optimization for  agentIdentifiers

Reply:
Oops, I forgot it. Yes, I agree. The `agentIdentifier` can be confusing, Maybe we can add a detailed explanation with an example in the help message? 

The ANP agent will validate the identifier. If the identifier is invalid, the agent will be registered as a random backend on the server-side. If the server can't find a backend by identifier, it will choose a random one, that's why even though the identifier is invalid here, but the server can still redirect the request.